### PR TITLE
[IOTDB-239] Add interface for showing devices

### PR DESCRIPTION
**Pull request by samperson1997**
Create an interface for showing all the devices.

USAGE: `SHOW DEVICES`.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/metadata/MTree.java [hunk 1]
```diff
@@ -760,6 +764,15 @@ private void findStorageGroup(MNode node, String path, HashSet<String> res) {
     }
   }
 
+  /**
+   * Get all devices in current Metadata Tree.
+   *
+   * @return a list contains all distinct storage groups
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MTree.java]:
```suggestion
   * @return a list contains all distinct devices
```

File: server/src/main/java/org/apache/iotdb/db/metadata/MTree.java [hunk 2]
```diff
@@ -760,6 +764,15 @@ private void findStorageGroup(MNode node, String path, HashSet<String> res) {
...
+   */
+  HashSet<String> getAllDevices() {
+    return new HashSet<>(getNodesList("3"));
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MTree.java]: why not using an int?

File: server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java [hunk 3]
```diff
@@ -311,15 +311,9 @@ public TSFetchMetadataResp fetchMetadata(TSFetchMetadataReq req) {
...
-          String column = req.getColumnPath();
-          Map<String, List<String>> deviceMap = metadata.getDeviceMap();
-          if (deviceMap == null || !deviceMap.containsKey(column)) {
-            resp.setColumnsList(new ArrayList<>());
-          } else {
-            resp.setColumnsList(deviceMap.get(column));
-          }
+        case "SHOW_DEVICES":
+          Set<String> devices = getAllDevices();
+          resp.setShowDevices(devices);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java]:
```suggestion
          resp.setDevices(devices);
```

File: server/src/main/java/org/apache/iotdb/db/metadata/MTree.java (hunk 2 above)
Comment (samperson1997) [reply server/src/main/java/org/apache/iotdb/db/metadata/MTree.java]: `getNodesList` method is implemented before and I just call it here, should I change the param type to int? (which I think is better as well)

File: server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java (hunk 3 above)
Comment (samperson1997) [reply server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java]: Here I also keep consistent with `SHOW TIMESERIES` and `SHOW STORAGE GROUP` interface, in which they uses `setShowTimeseriesList` and `setShowStorageGroups`... Should I change all of these interfaces? (and I also think your suggestion is better...)

File: jdbc/src/test/java/org/apache/iotdb/jdbc/IoTDBDatabaseMetadataTest.java [hunk 4]
```diff
@@ -240,7 +240,41 @@ public void device() throws Exception {
...
+      resultStr.append("\n");
+      while (resultSet.next()) {
+        for (int i = 1; i <= colCount; i++) {
+          resultStr.append(resultSet.getString(i)).append(",");
+        }
+        resultStr.append("\n");
+      }
+      Assert.assertEquals(resultStr.toString(), standard);
+    } catch (SQLException e) {
+      System.out.println(e);
```
Comment (kr11) [inline jdbc/src/test/java/org/apache/iotdb/jdbc/IoTDBDatabaseMetadataTest.java]: Simply  print out the exception may bury the error and the TestCase will be regarded as `PASSED`. It's better to replace it with `e.printStackTrace(); fail(e.getMessage());`.

File: server/src/main/java/org/apache/iotdb/db/metadata/MGraph.java [hunk 5]
```diff
@@ -248,6 +248,10 @@ public Metadata getMetadata() throws PathErrorException {
     return mtree.getAllStorageGroup();
   }
 
+  HashSet<String> getAllDevices() {
```
Comment (kr11) [inline server/src/main/java/org/apache/iotdb/db/metadata/MGraph.java]: I note that all its callers Interprete the returned value as `Set` but not `HashSet`.

**Comment by samperson1997** [comment]
Hi @qiaojialin @kr11 ,

Thanks a lot for your detailed review! I have updated codes as you suggested.

**Pull request merged**